stl/inc/algorithm

```diff
@@ -3342,13 +3366,242 @@ _FwdIt2 remove_copy_if(_ExPo&&, _FwdIt1 _First, _FwdIt1 _Last, _FwdIt2 _Dest, _P
+
+    inline constexpr _Remove_fn remove{_Not_quite_object::_Construct_tag{}};
+
+    // VARIABLE ranges::remove_if
+    class _Remove_if_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        template <permutable _It, sentinel_for<_It> _Se, class _Pj = identity,
+            indirect_unary_predicate<projected<_It, _Pj>> _Pr>
+        constexpr subrange<_It> operator()(_It _First, _Se _Last, _Pr _Pred, _Pj _Proj = {}) const {
+            _Adl_verify_range(_First, _Last);
+            auto _UResult = _Remove_if_unchecked(
+                _Get_unwrapped(_STD move(_First)), _Get_unwrapped(_STD move(_Last)), _Pass_fn(_Pred), _Pass_fn(_Proj));
+
+            return _Rewrap_subrange<subrange<_It>>(_First, _STD move(_UResult));
+        }
+
+        // clang-format off
+        template <forward_range _Rng, class _Pj = identity,
+            indirect_unary_predicate<projected<iterator_t<_Rng>, _Pj>> _Pr>
+            requires permutable<iterator_t<_Rng>>
+        constexpr borrowed_subrange_t<_Rng> operator()(_Rng&& _Range, _Pr _Pred, _Pj _Proj = {}) const {
+            auto _UResult = _Remove_if_unchecked(_Ubegin(_Range), _Uend(_Range), _Pass_fn(_Pred), _Pass_fn(_Proj));
+
+            return _Rewrap_subrange<borrowed_subrange_t<_Rng>>(_Range, _STD move(_UResult));
+        }
+        // clang-format on
+    private:
+        template <class _It, class _Se, class _Pr, class _Pj>
+        _NODISCARD static constexpr subrange<_It> _Remove_if_unchecked(
+            _It _First, const _Se _Last, _Pr _Pred, _Pj _Proj) {
+            // Remove values whose projection satisfies _Pred from [_First, _Last)
+            _STL_INTERNAL_STATIC_ASSERT(permutable<_It>);
+            _STL_INTERNAL_STATIC_ASSERT(sentinel_for<_Se, _It>);
+            _STL_INTERNAL_STATIC_ASSERT(indirect_unary_predicate<_Pr, projected<_It, _Pj>>);
+
+            _First     = _RANGES _Find_if_unchecked(_STD move(_First), _Last, _Pred, _Proj);
+            auto _Next = _First;
+            if (_First == _Last) {
+                return {_STD move(_Next), _STD move(_First)};
+            }
+
+            while (++_First != _Last) {
+                if (!_STD invoke(_Pred, _STD invoke(_Proj, *_First))) {
+                    *_Next = _RANGES iter_move(_First);
+                    ++_Next;
+                }
+            }
+
+            return {_STD move(_Next), _STD move(_First)};
+        }
+    };
+
+    inline constexpr _Remove_if_fn remove_if{_Not_quite_object::_Construct_tag{}};
+
+    // ALIAS TEMPLATE remove_copy_result
+    template <class _In, class _Out>
+    using remove_copy_result = in_out_result<_In, _Out>;
+
+    // VARIABLE ranges::remove_copy
+    class _Remove_copy_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        // clang-format off
+        template <permutable _It, sentinel_for<_It> _Se, weakly_incrementable _Out, class _Ty, class _Pj = identity>

```

Comment:
`remove` and `remove_if` require `permutable` because they mutate the source range in-place; `remove_copy` and `remove_copy_if` don't because they copy (a subset of) elements from the input into their output.
```suggestion
        template <input_iterator _It, sentinel_for<_It> _Se, weakly_incrementable _Out, class _Ty, class _Pj = identity>
```

Reply:
(Don't forget to change the iterator un-/re-wrapping consistently.)

Reply:
Argh, it seems my brain did not want to deal with input ranges anymore